Allow to customize track preparation

danielmitterdorfer opened the pull request:
With this commit we introduce track processors and define two phases in the
track lifecycle where they can hook into:

* `on_after_load_track`: This callback is executed after the track has been
successfully loaded and converted into Rally's domain representation. Rally
defines some out-of-the-box track processors for task filters and test mode
support which will be called in any case.
* `on_prepare_track`: This callback is executed before a load generator will
start executing the benchmark. It can be used to download or generate any
datasets that are needed for the benchmark. Rally defines a default behavior
that downloads and extracts all relevant corpora. That behavior can be
overridden via a custom track processor (i.e. Rally's default behavior is then
*disabled*.).

Like all custom track components, track processors are registered via the
`register` function and we implement a new registration method on the registry
called `register_track_processor` for that, which expects an instance that
implements the two methods above.

Note that track processors cannot share state in between these phases. Some
phases might only be invoked on the coordinating node of a benchmark
(`on_after_load_track`) whereas others will need to be called on each load
generator machine (`on_prepare_track`).

We introduce a new `parameters` dictionary on track as well as on challenge
level, where the challenge properties override the defaults defined track level.
These parameters are only exposed on challenge level and track processors are
expected to operate only on that level. Track processors can determine the
current challenge with the new property `selected_challenge_or_default`.
Mid-term we intend to restrict the current track parameter feature in a way that
all values provided via `--track-params` on the command line must be declared
within the `parameters` block, either on track or on challenge level. That's the
reason why we have called this block already `parameters` although it will only
be used by track processors for now.

We intentionally do not provide user documentation for this feature yet as we
want to gain more experience first before exposing it to a wider user base.

Closes #1066

dliappis commented on esrally/utils/collections.py:
```diff
@@ -0,0 +1,39 @@
+from typing import Any, Generator, Mapping
+
+
+def merge_dicts(d1: Mapping[str, Any], d2: Mapping[str, Any]) -> Generator[Any, None, Any]:
```
An alternative non-recursive implementation that seems to be quite speedy can be found here: https://stackoverflow.com/a/61708681/12047396, however, it won't merge lists.

danielmitterdorfer replied:
I'd go for the more generic implementation as I don't think performance matters too much here.

dliappis reviewed (approved):
LGTM! Left a few comments and questions, but the implementation looks solid.

danielmitterdorfer commented on esrally/track/loader.py:
```diff
@@ -41,6 +41,59 @@ class TrackSyntaxError(exceptions.InvalidSyntax):
+
+        :param track: The current track.
+        """
+        ...
```
I'll change to `pass`

danielmitterdorfer commented on esrally/track/loader.py:
```diff
@@ -692,120 +792,137 @@ def relative_glob(start, f):
+                        path, ext = io.splitext(document_set.document_archive)
+                        path_2, ext_2 = io.splitext(path)
+
+                        document_set.document_archive = "%s-1k%s%s" % (path_2, ext_2, ext)
```
No, this was mainly because I was only moving lines around but did not change the implementation. I'll adapt accordingly.

danielmitterdorfer commented:
Thanks for your review @dliappis. I've addressed all your review comments in https://github.com/elastic/rally/pull/1135/commits/9d752bbc3de95d7063d8cf6a3b0c606cf29dbbb3. I let CI run once more and will merge it then.

danielmitterdorfer commented:
@elasticmachine test this please
